File: packages/shared/src/cognitive/paper-daily-report.ts
/**
 * V1.46/V1.47 — Paper Daily Report / autoDesk (proyección de PaperDeskCycle).
 *
 * @see docs/engineering/spec-v147-paper-desk-runtime-truth-2026-09-01.md
 */

import type { PaperDeskNextActionV1 } from "./operational-context.js";
import type { TradePlanV1 } from "./trade-plan.js";

export const PAPER_DAILY_REPORT_SCHEMA = "paper_daily_report_v1" as const;

/** V1.50 — reason codes de EntryTick (mirror Python EntryReasonCode). */
export type PaperDeskEntryReasonCodeV1 =
  | "ENTRY_NO_TRIGGER"
  | "ENTRY_INVALID_STOP"
  | "ENTRY_RISK_LIMIT"
  | "ENTRY_STALE_DATA"
  | "ENTRY_MANDATE_BLOCK"
  | "ENTRY_POLICY_MISSING"
  | "ENTRY_MARKET_CLOSED"
  | "ENTRY_DUPLICATE"
  | "ENTRY_ENV_BLOCKED"
  | "ENTRY_UNIVERSE_EMPTY"
  | "ENTRY_UNIVERSE_UNAVAILABLE"
  | "ENTRY_INFRA_UNAVAILABLE";

/** V1.50 — CandidateSnapshot transport (mirror Python paper_desk_cycle). */
export type PaperDeskCandidateSnapshotV1 = {
  decisionId: string;
  instrumentId: string;
  rank: number;
  score: number;
  symbol?: string | null;
  autoSource?: string | null;
  templateId?: string | null;
  analysisAsOf?: string | null;
  marketAsOf?: string | null;
  executionAsOf?: string | null;
  tradePlan?: TradePlanV1 | null;
  entry?: number | null;
  structuralStop?: number | null;
  target1?: number | null;
  target2?: number | null;
  riskAmount?: number | null;
  expectedRr?: number | null;
  mandate?: string;
  freshness?: string;
  vetoes?: string[];
  reasonCode?: PaperDeskEntryReasonCodeV1 | null;
  humanMessage?: string | null;
};

/** V1.54 — excepciones operativas para cubo 🔴 (sin auto-heal). */
export type DailyDeskExceptionKindV1 =
  | "position_birth_failed"
  | "portfolio_recon_drift"
  | "portfolio_recon_unavailable";

export type DailyDeskExceptionFactV1 = {
  kind: DailyDeskExceptionKindV1;
  instrumentId?: string | null;
  symbol?: string | null;
  decisionId?: string | null;
  message?: string | null;
};

export type PaperDailyReportEntryV1 = {
  status: string;
  proposed: number;
  executed: number;
  candidates?: PaperDeskCandidateSnapshotV1[];
  skipped?: PaperDeskCandidateSnapshotV1[];
};

export type PaperDailyReportJitDenyCodeV1 =
  | "data_stale"
  | "market_closed"
  | "portfolio_drift"
  | "paper_auto_env_blocked"
  | "data_unavailable";

export type PaperDeskPositionRowV1 = {
  instrumentId: string;
  status: string;
  reason?: string | null;
  decisionVerdict?: string | null;
  permissionReasons?: string[];
  nextAction?: PaperDeskNextActionV1 | string | null;
  decisionAction?: string | null;
  executedAction?: string | null;
  operatingState?: string | null;
};

export type PaperDailyReportSectionsV1 = {
  decisiones: {
    candidates: number;
    proposed: number;
    authorized: number;
    executed: number;
  };
  operativa: {
    entries: number;
    t1: number;
    trails: number;
    exits: number;
  };
  resultado: {
    realizedR: number | null;
    dayPct: number | null;
  };
  noOperadas: {
    skipped: number;
    reasonCodes: Partial<Record<PaperDeskEntryReasonCodeV1, number>>;
  };
};

export type PaperDailyReportV1 = {
  schemaVersion: typeof PAPER_DAILY_REPORT_SCHEMA;
  accountId: string;
  asOf: string | null;
  dryRun: boolean;
  paperDExecute: boolean;
  blocked?: boolean;
  blockReason?: string | null;
  entry: PaperDailyReportEntryV1;
  /** V1.55 — DECISIONES / OPERATIVA / RESULTADO / NO OPERADAS */
  sections?: PaperDailyReportSectionsV1;
  /** V1.54 — hechos de excepción para proyección Desk (opcional). */
  exceptionFacts?: DailyDeskExceptionFactV1[];
  positions: {
    held: number;
    denied: number;
    protected: number;
    reduced: number;
    exited: number;
    rows: PaperDeskPositionRowV1[];
  };
  jitDenies: Partial<Record<PaperDailyReportJitDenyCodeV1, number>>;
  notes: string[];
};

export type PaperDeskCycleLikeV1 = {
  accountId: string;
  asOf?: string | null;
  dryRun: boolean;
  paperDExecute: boolean;
  blocked?: boolean;
  blockReason?: string | null;
  entry: {
    status: string;
    proposedCount?: number;
    executedCount?: number;
    notes?: string[];
  };
  positions: PaperDeskPositionRowV1[];
  notes?: string[];
};

const JIT_CODES: readonly PaperDailyReportJitDenyCodeV1[] = [
  "data_stale",
  "market_closed",
  "portfolio_drift",
  "paper_auto_env_blocked",
  "data_unavailable",
] as const;

export function buildPaperDailyReport(
  cycle: PaperDeskCycleLikeV1,
): PaperDailyReportV1 {
  const jitDenies: Partial<Record<PaperDailyReportJitDenyCodeV1, number>> = {};
  for (const code of JIT_CODES) {
    jitDenies[code] = 0;
  }

  const notes = [...(cycle.notes ?? []), ...(cycle.entry.notes ?? [])];
  if (cycle.blocked || cycle.blockReason === "paper_auto_env_blocked") {
    jitDenies.paper_auto_env_blocked =
      (jitDenies.paper_auto_env_blocked ?? 0) + 1;
    if (!notes.includes("paper_auto_env_blocked")) {
      notes.push("paper_auto_env_blocked");
    }
  }
  if (cycle.dryRun && !notes.some((n) => n.toLowerCase().includes("dryrun"))) {
    notes.push("dryRun");
  }

  let held = 0;
  let denied = 0;
  let protectedCount = 0;
  let reduced = 0;
  let exited = 0;

  for (const row of cycle.positions) {
    if (row.status === "held") held += 1;
    else if (row.status === "denied") {
      denied += 1;
      for (const code of row.permissionReasons ?? []) {
        if ((JIT_CODES as readonly string[]).includes(code)) {
          const key = code as PaperDailyReportJitDenyCodeV1;
          jitDenies[key] = (jitDenies[key] ?? 0) + 1;
        }
      }
      if (row.reason?.includes("paper_auto_env_blocked")) {
        jitDenies.paper_auto_env_blocked =
          (jitDenies.paper_auto_env_blocked ?? 0) + 1;
      }
    } else if (row.status === "protected") protectedCount += 1;
    else if (row.status === "reduced") reduced += 1;
    else if (row.status === "exited") exited += 1;
  }

  return {
    schemaVersion: PAPER_DAILY_REPORT_SCHEMA,
    accountId: cycle.accountId,
    asOf: cycle.asOf ?? null,
    dryRun: cycle.dryRun,
    paperDExecute: cycle.paperDExecute,
    blocked: cycle.blocked,
    blockReason: cycle.blockReason ?? null,
    entry: {
      status: cycle.entry.status,
      proposed: cycle.entry.proposedCount ?? 0,
      executed: cycle.entry.executedCount ?? 0,
    },
    sections: buildPaperDailyReportSections(cycle, {
      protected: protectedCount,
      reduced,
      exited,
    }),
    positions: {
      held,
      denied,
      protected: protectedCount,
      reduced,
      exited,
      rows: [...cycle.positions],
    },
    jitDenies,
    notes,
  };
}

export function buildPaperDailyReportSections(
  cycle: PaperDeskCycleLikeV1,
  counts: { protected: number; reduced: number; exited: number },
): PaperDailyReportSectionsV1 {
  const proposed = cycle.entry.proposedCount ?? 0;
  const executed = cycle.entry.executedCount ?? 0;
  const skippedReasons: Partial<Record<PaperDeskEntryReasonCodeV1, number>> =
    {};
  let trails = 0;
  for (const row of cycle.positions) {
    if (row.decisionAction === "TRAIL" || row.decisionVerdict === "TRAIL") {
      trails += 1;
    }
  }
  return {
    decisiones: {
      candidates: proposed,
      proposed,
      authorized: executed > 0 ? 1 : 0,
      executed,
    },
    operativa: {
      entries: executed,
      t1: counts.reduced,
      trails,
      exits: counts.exited,
    },
    resultado: {
      realizedR: null,
      dayPct: null,
    },
    noOperadas: {
      skipped: Math.max(0, proposed - executed),
      reasonCodes: skippedReasons,
    },
  };
}

File: packages/py/application/src/bolsa_application/paper_daily_report.py
"""V1.46 — PaperDailyReport / autoDesk projection from PaperDeskCycleResult."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from bolsa_application.paper_desk_cycle import PaperDeskCycleResult

PAPER_DAILY_REPORT_SCHEMA = "paper_daily_report_v1"

_JIT_DENY_CODES = frozenset(
    {
        "data_stale",
        "market_closed",
        "portfolio_drift",
        "paper_auto_env_blocked",
        "data_unavailable",
    }
)


@dataclass
class PaperDailyReportV1:
    schema_version: str = PAPER_DAILY_REPORT_SCHEMA
    account_id: str = ""
    as_of: str | None = None
    dry_run: bool = True
    paper_d_execute: bool = False
    entry_proposed: int = 0
    entry_executed: int = 0
    entry_status: str = "skipped"
    entry_candidates: list[dict[str, Any]] = field(default_factory=list)
    entry_skipped: list[dict[str, Any]] = field(default_factory=list)
    exception_facts: list[dict[str, Any]] = field(default_factory=list)
    position_held: int = 0
    position_denied: int = 0
    position_protected: int = 0
    position_reduced: int = 0
    position_exited: int = 0
    position_rows: list[dict[str, Any]] = field(default_factory=list)
    jit_denies: dict[str, int] = field(default_factory=dict)
    notes: list[str] = field(default_factory=list)
    blocked: bool = False
    block_reason: str | None = None

    def to_dict(self) -> dict[str, Any]:
        entry: dict[str, Any] = {
            "status": self.entry_status,
            "proposed": self.entry_proposed,
            "executed": self.entry_executed,
        }
        if self.entry_candidates:
            entry["candidates"] = list(self.entry_candidates)
        if self.entry_skipped:
            entry["skipped"] = list(self.entry_skipped)

        out: dict[str, Any] = {
            "schemaVersion": self.schema_version,
            "accountId": self.account_id,
            "asOf": self.as_of,
            "dryRun": self.dry_run,
            "paperDExecute": self.paper_d_execute,
            "blocked": self.blocked,
            "blockReason": self.block_reason,
            "entry": entry,
            "positions": {
                "held": self.position_held,
                "denied": self.position_denied,
                "protected": self.position_protected,
                "reduced": self.position_reduced,
                "exited": self.position_exited,
                "rows": list(self.position_rows),
            },
            "jitDenies": dict(self.jit_denies),
            "notes": list(self.notes),
        }
        if self.exception_facts:
            out["exceptionFacts"] = list(self.exception_facts)
        sections = _build_report_sections(self)
        if sections:
            out["sections"] = sections
        return out


def _build_report_sections(report: PaperDailyReportV1) -> dict[str, Any]:
    """V1.55 — DECISIONES / OPERATIVA / RESULTADO / NO OPERADAS."""
    trails = sum(
        1
        for row in report.position_rows
        if row.get("decisionAction") == "TRAIL"
        or row.get("decisionVerdict") == "TRAIL"
    )
    proposed = report.entry_proposed
    executed = report.entry_executed
    return {
        "decisiones": {
            "candidates": proposed,
            "proposed": proposed,
            "authorized": 1 if executed > 0 else 0,
            "executed": executed,
        },
        "operativa": {
            "entries": executed,
            "t1": report.position_reduced,
            "trails": trails,
            "exits": report.position_exited,
        },
        "resultado": {
            "realizedR": None,
            "dayPct": None,
        },
        "noOperadas": {
            "skipped": max(0, proposed - executed),
            "reasonCodes": {},
        },
    }


def _collect_exception_facts(cycle: PaperDeskCycleResult) -> list[dict[str, Any]]:
    """V1.54 — hechos operativos para proyección Desk (notes + entry)."""
    facts: list[dict[str, Any]] = []
    seen: set[str] = set()
    all_notes = list(cycle.notes) + list(cycle.entry.notes)

    def _add(kind: str, **fields: Any) -> None:
        if kind in seen:
            return
        seen.add(kind)
        fact: dict[str, Any] = {"kind": kind}
        for key, value in fields.items():
            if value is not None:
                fact[key] = value
        facts.append(fact)

    birth_note = any("position_birth_failed" in n for n in all_notes)
    birth_reason = cycle.entry.reason == "position_birth_failed"
    if birth_note or birth_reason:
        ctx: dict[str, Any] = {}
        for cand in cycle.entry.candidates:
            ctx = {
                "instrumentId": cand.instrument_id,
                "symbol": cand.symbol,
                "decisionId": cand.decision_id,
            }
            break
        _add("position_birth_failed", **ctx)

    if any(n == "portfolio_drift" or "portfolio_drift" in n for n in all_notes):
        _add("portfolio_recon_drift")

    if any(n == "recon_unavailable" or "recon_unavailable" in n for n in all_notes):
        _add("portfolio_recon_unavailable")

    return facts


def build_paper_daily_report(cycle: PaperDeskCycleResult) -> PaperDailyReportV1:
    """Proyecta PaperDeskCycleResult → PaperDailyReportV1 (autoDesk)."""
    jit: dict[str, int] = {k: 0 for k in sorted(_JIT_DENY_CODES)}
    notes = list(cycle.notes)
    notes.extend(cycle.entry.notes)

    if cycle.blocked or cycle.block_reason == "paper_auto_env_blocked":
        jit["paper_auto_env_blocked"] = jit.get("paper_auto_env_blocked", 0) + 1
        if "paper_auto_env_blocked" not in notes:
            notes.append("paper_auto_env_blocked")

    if cycle.dry_run and "dryRun" not in "".join(notes):
        notes.append("dryRun")

    held = denied = protected = reduced = exited = 0
    rows: list[dict[str, Any]] = []
    for p in cycle.positions:
        rows.append(p.to_dict())
        if p.status == "held":
            held += 1
        elif p.status == "denied":
            denied += 1
            for code in p.permission_reasons:
                if code in _JIT_DENY_CODES:
                    jit[code] = jit.get(code, 0) + 1
            if p.reason and "paper_auto_env_blocked" in p.reason:
                jit["paper_auto_env_blocked"] = jit.get("paper_auto_env_blocked", 0) + 1
        elif p.status == "protected":
            protected += 1
        elif p.status == "reduced":
            reduced += 1
        elif p.status == "exited":
            exited += 1

    return PaperDailyReportV1(
        account_id=cycle.account_id,
        as_of=cycle.as_of,
        dry_run=cycle.dry_run,
        paper_d_execute=cycle.paper_d_execute,
        entry_proposed=cycle.entry.proposed_count,
        entry_executed=cycle.entry.executed_count,
        entry_status=cycle.entry.status,
        entry_candidates=[c.to_dict() for c in cycle.entry.candidates],
        entry_skipped=[s.to_dict() for s in cycle.entry.skipped],
        exception_facts=_collect_exception_facts(cycle),
        position_held=held,
        position_denied=denied,
        position_protected=protected,
        position_reduced=reduced,
        position_exited=exited,
        position_rows=rows,
        jit_denies=jit,
        notes=notes,
        blocked=cycle.blocked,
        block_reason=cycle.block_reason,
    )
